File: Python3/1 - 200/84 Largest Rectangle in Histogram.py
# 68ms, 66.48%; 14.6MB, 33.54%
class Solution:
    def largestRectangleArea(self, heights: list) -> int:
        heights.append(0)
        res, stack = 0, [] 
        i, size = 0, len(heights)
        while i < size:
            if not stack or heights[i] > heights[stack[-1]]:
                stack.append(i)
                i += 1
            else:
                tmp = stack.pop()
                res = max(res, heights[tmp] * (i if not stack else i - stack[-1] - 1))                
        return res

File: 1 - 200/84. Largest Rectangle in Histogram.cc
// 8ms, 100.0%; 10.8MB, 20.35%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    int largestRectangleArea(vector<int> &heights) {
        heights.push_back(0);
        stack<int> st;
        int res = 0, i = 0, size = heights.size();
        while (i < size) {
            if (st.empty() || heights[i] > heights[st.top()]) {
                st.push(i++);
            } else {
                int ptr = st.top();
                st.pop();
                res = max(res, heights[ptr] * (st.empty() ? i : i - st.top() - 1));
            }
        }
        return res;
    }
};
